### ISSUE TITLE
Batch normalization NaN on GPU with optimizer

### ISSUE THREAD

**Issue by lukas-schott** [Theano/Theano#4368]
Hey,
if I run a class lasagne.layers.BatchNormLayer layer for the input data. I could reduce my error to the following piece of code (should run plug and play):

```
import lasagne
import numpy as np
import theano
from theano.sandbox import cuda as s

# genrates a matrix of shape (batch_size, sequ_len, n_feat) with one hot features at t=0.
def generate_rand_data(batch_size, sequ_len, n_feat):
    rand_inds = np.random.randint(n_feat, size=batch_size)
    batch = np.zeros((batch_size, sequ_len, n_feat), dtype=theano.config.floatX)
    batch[range(batch_size), 0, rand_inds] = 1
    return batch


if __name__ == "__main__":
    batch_size, sequ_len, n_feat = 2, 3, 4

    l_in = lasagne.layers.InputLayer((batch_size, sequ_len, n_feat))
    l_bnorm = lasagne.layers.BatchNormLayer(l_in, axes=(0)) # normalize each input at each t

    debug_2 = theano.function([l_in.input_var],
                              lasagne.layers.get_output(l_bnorm))

    batch = generate_rand_data(batch_size, sequ_len, n_feat)

    print 'input', batch
    print 'norm layer', debug_2(batch)
```

I get the following NaN output with THEANO_FLAGS='device=gpu' python test.py:

```
input [[[ 1.  0.  0.  0.]
  [ 0.  0.  0.  0.]
  [ 0.  0.  0.  0.]]

 [[ 0.  0.  1.  0.]
  [ 0.  0.  0.  0.]
  [ 0.  0.  0.  0.]]]

norm layer [[[ nan   0.  nan   0.]    # should be 0.9998 and -0.9998 instead of NaNs
  [  0.   0.   0.   0.]
  [  0.   0.   0.   0.]]

 [[ nan   0.  nan   0.]  # should be -0.9998 and 0.9998 instead of NaNs
  [  0.   0.   0.   0.]
  [  0.   0.   0.   0.]]
```

If I run it on the CPU or with THEANO_FLAGS='device=gpu, optimizer=None' python test.py it works

**Comment by nouiz** [comment]
if you use device=gpu and optimizer=None, it won't be on the GPU.

When on CPU, do you have floatX=float32 as on GPU?

Can you try the Theano flag mode=DebugMode to know exactly which optimization introduce this?

**Comment by yshean** [comment]
I experience the same issue when running on GPU, but no problem at all on CPU. I had to switch to Tensorflow to utilize GPU. 

@llioy Have you managed to solve the issue?

**Comment by yshean** [comment]
I managed to overcome this issue by setting nvcc.fastmath=False in THEANO_FLAGS, as suggested here: http://deeplearning.net/software/theano/tutorial/nan_tutorial.html

**Comment by llioy** [comment]
@nouiz : Yeah, I had floatX=float32 (as default in the theanorc). "Unfortunately" Im now not able to reproduce the problem with the most recent Lasgne and Theano versions. So no more unextpected NaN problems ever since then!

@yshean it solved itself with the most recent Lasagne and Theano version (commit b5018db) and Lasagne (commit 8b66737).

**Comment by nouiz** [comment]
Ok. So closing it for now.
